### DISCUSSION TITLE
CorruptComponentPayloadException with built-in WithPagination trait

### DISCUSSION THREAD

**Discussion by dennisprudlo** [livewire/livewire#5352] [7. Bugs (archived)]
Hey all,

I have been struggling with an annoying issue I had with my listings and was able to solve it now. Turns out that the `WithPagination` was the source of it all :D

### What's the issue?
I have a very common listing of paginated items with some filters to apply. Navigating through the pages and applying filters work without an issue, but heres the thing: When I navigate to say page 4, then click on an item (href to a "details page") and then use the browsers "Back button" functionality to return to the index, livewire shows the correct items. But as soon as I change anything resulting in a livewire request, the `CorruptComponentPayloadException` exception is thrown.

I thought this is some issue with my filters or other custom properties but the same scenario works on the first page. Click on item (href to details page), use the back button to return, then updating the search field or other stuff. There is no issue with the first page.

### Where's the problem?
After some digging I found out that the `WithPagination` trait itself relies on integers for the page numbers (which makes sense). But it does not convert the values to integers when reading from the query string. So on the first page the page number is either `1` or not present (because the page number is being removed from the query string when it's the first page) which will default to `1`. However when reading from the query string the page number is not an integer but a string. So when navigating to page 4, livewire sets the property to `4` and adds `page=4` to the query string. Clicking on a link and then returning to the page will read the query string and set the page to `"4"`. This "inconsistency" will result in a different checksum and livewire rejects the request with the corrupt payload exception.

There are situations in other similar components where the `"4"` is kept as a string and the pagination works with that. But I think the best solution is to always use the same type for page numbers in the paginator. These logs show the `fingerprint`, `memo` and `checksum` *just before* hydration and *just after* dehydration:

<img width="1305" alt="Screenshot 2022-11-08 at 6 20 55 PM" src="https://user-images.githubusercontent.com/50237950/200632610-13645234-074c-44e5-8534-78462bc0a305.png">

You can see how in `memo.data` the type of the `page` value changes from integer to string and back to integer. It seems to work with the most simple way of using pagination in livewire, however with more complex components that add lazy loading to each paginated page, etc. these casting issues happen.

### What's the solution?
The I guess least invasive and most effective solution would be to simply enforce a type for the page numbers. Unfortunately there is no way to hook into the `WithPaginator` trait to do it elegantly for multiple paginators. A simple fix would be to cast the page values to integers every time we assign them to a variable – just like the `setPage(...)` method is already doing. The following changes in the `WithPaginator` trait resolved the issue in my case:

<img width="1684" alt="Screenshot 2022-11-08 at 6 30 59 PM" src="https://user-images.githubusercontent.com/50237950/200634742-4b212050-1b96-4d3e-9ed7-d675a0ef7e49.png">

I don't know about cursor pagination and if casting a possible empty string to a `0` is a valid solution in that case. I am also not sure if with upcoming Livewire 3 this is something to be fixed here or if I should simply copy the trait and maintain a local version of it.

Would love the hear some feedback on this and if it's worth looking into.

Thanks!

**Comment by joshhanley** [comment]
@dennisprudlo ah interesting, thanks for reporting! I will have a look when I get the chance and see if I can replicate it 🙂
